pkg/binary/envoy/debug/pii.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2019 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug
+
+import (
+	"crypto/sha256"
+	"fmt"
+
+	"bitbucket.org/creachadair/shell"
+)
+
+// defaultHash returns the hashed value of s using sha256 defaultHash function
+// TODO: salt the Hash
+//nolint: unused,deadcode
+func defaultHash(s string) string {
+	h := sha256.New()
+	_, _ = h.Write([]byte(s))
+	return string(h.Sum(nil))
+}
+
+var defaultFormat = `[%START_TIME%] "%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%"` +
+	` %RESPONSE_CODE% %RESPONSE_FLAGS% "%DYNAMIC_METADATA(istio.mixer:status)%" "%REQ(USER-AGENT)%"`
+var defaultPII = map[string]bool{
+	"[%START_TIME%]": true,
+}
+
+// Filter filters log fields using pii and modify all PII fields using f
+type Filter struct {
+	f      func(string) string
+	pii    map[string]bool
+	format []string
+}
+
+// NewFilter constructs a custom filter object
+func NewFilter(formatStr string, f func(string) string, pii map[string]bool) (Filter, error) {
+	// splitting formats and handle error
+	format, ok := shell.Split(formatStr)
+	if !ok {
+		return Filter{}, fmt.Errorf("error in splitting format string: %s", format)
+	}
+	return Filter{f: f, pii: pii, format: format}, nil
+}
+
+// process logs with f the filter, assumes that filter has valid fields
+func (filter Filter) process(logs []string) []string {
+	out := make([]string, 0, len(logs))
+	for _, log := range logs {
+		fieldValues, ok := shell.Split(log)
+		if !ok {
+			fmt.Printf("error in splitting log: %s", log)
+			continue
```

Comment:
IMO: In this PII filtering area, we should filter/mask all unknown/invalid-format values rather than leaving them unprocessed to avoid accidental PII logging.

Reply:
We omit the log entirely by calling `continue` here - so if we don't know we won't just hash the entire log line, we'll ignore it entirely and it won't appear in output.

Agree we _probably_ shouldn't `fmt.Printf` the log line itself (since it can contain PII)